util/rateLimit.go

```diff
@@ -0,0 +1,470 @@
+
+	"github.com/algorand/go-algorand/util/metrics"
+	"github.com/algorand/go-deadlock"
+)
+
+// ElasticRateLimiter holds and distributes capacity through capacityQueues
+// Capacity consumers are given an error if there is no capacity available for them,
+// and a "capacityGuard" structure they can use to return the capacity when finished
+type ElasticRateLimiter struct {
+	MaxCapacity            int
+	CapacityPerReservation int
+	sharedCapacity         capacityQueue
+	capacityByClient       map[ErlClient]capacityQueue
+	clientLock             deadlock.RWMutex
+	noCapacityCounter      *metrics.Counter
+	// CongestionManager and enable flag
+	cm                       CongestionManager
+	enableCM                 bool
+	congestionControlCounter *metrics.Counter
+}
+
+// ErlClient clients must support OnClose for reservation closing
+type ErlClient interface {
+	OnClose(func())
+}
+
+// capacity is an empty structure used for loading and draining queues
+type capacity struct {
+}
+
+// Capacity Queue wraps and maintains a channel of opaque capacity structs
+type capacityQueue chan capacity
+
+// ErlCapacityGuard is the structure returned to clients so they can release the capacity when needed
+// they also inform the congestion manager of events
+type ErlCapacityGuard struct {
+	client ErlClient
+	cq     capacityQueue
+	cm     *CongestionManager
+}
+
+// Release will put capacity back into the queue attached to this capacity guard
+func (cg ErlCapacityGuard) Release() error {
+	select {
+	case cg.cq <- capacity{}:
+		return nil
+	default:
+		return fmt.Errorf("could not replace capacity to channel: %v", cg.cq)
+	}
+}
+
+// Served will notify the CongestionManager that this resource has been served, informing the Service Rate
+func (cg ErlCapacityGuard) Served() {
+	if *cg.cm != nil {
+		(*cg.cm).Served(time.Now())
+	}
+}
+
+func (q capacityQueue) blockingRelease() {
+	q <- capacity{}
+}
+
+func (q capacityQueue) blockingConsume() {
+	<-q
+}
+
+func (q capacityQueue) consume(cm *CongestionManager) (ErlCapacityGuard, error) {
+	select {
+	case <-q:
+		return ErlCapacityGuard{
+			cq: q,
+			cm: cm,
+		}, nil
+	default:
+		return ErlCapacityGuard{}, fmt.Errorf("could not consume capacity from capacityQueue: %v", q)
+	}
+}
+
+// NewElasticRateLimiter creates an ElasticRateLimiter and initializes maps
+func NewElasticRateLimiter(
+	maxCapacity int,
+	reservedCapacity int,
+	cm CongestionManager,
+	nocapCount *metrics.Counter,
+	conmanCount *metrics.Counter) *ElasticRateLimiter {
+	ret := ElasticRateLimiter{
+		MaxCapacity:              maxCapacity,
+		CapacityPerReservation:   reservedCapacity,
+		capacityByClient:         map[ErlClient]capacityQueue{},
+		cm:                       cm,
+		sharedCapacity:           capacityQueue(make(chan capacity, maxCapacity)),
+		noCapacityCounter:        nocapCount,
+		congestionControlCounter: conmanCount,
+	}
+	// fill the sharedCapacity
+	for i := 0; i < maxCapacity; i++ {
+		ret.sharedCapacity.blockingRelease()
```

Comment:
enqueueing 60k elements into a chan/queue on startup looks interesting. Have considered other approaches to implement shared capacity like, say as a atomic counter with a cap?

Reply:
we did talk about that, but wanted to use channels because they naturally behave like queues, and would not require any coordination when `releasing` the capacity.

Is the concern just that this is potentially costly at startup? If so, how can we measure that?

The ERL is designed to never over-provision capacity, so the ~60k units created should not grow over time.